File: examples/x3_oop_vehicles.py
#!/usr/bin/env python3
'''
@author Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d
from x2_oop_vehicle import Vehicle

def tick():
    g2d.clear_canvas()
    for a in actors:
        a.move()
        g2d.fill_rect(a.position())

actors = [Vehicle((40, 40), 5),
          Vehicle((120, 40), 5),
          Vehicle((80, 80), -5)]
g2d.init_canvas((480, 360))
g2d.main_loop(tick)

File: go/x3_oop_vehicles.go
package main

import . "g2d"

var screen = Point{480, 360}
var margin = 100

var actors = []*Vehicle { NewVehicle(Point{40, 40}, 5),
    NewVehicle(Point{80, 80}, -5), NewVehicle(Point{120, 40}, 5) }

type Vehicle struct {
    x, y, w, h      int
    dx, left, right int
}

func NewVehicle(pos Point, dx int) *Vehicle {
    return &Vehicle{pos.X, pos.Y, 20, 20, dx, -margin, screen.X+margin}
}

func (a *Vehicle) Move() {
    if a.x + a.dx < a.left {
        a.x = a.right
    }
    if a.x + a.dx > a.right {
        a.x = a.left
    }
    a.x += a.dx
}

func (a *Vehicle) Position() Point {
    return Point{a.x, a.y}
}

func tick() {
    ClearCanvas()
    for _, a := range actors {
        a.Move()
        DrawImage("ball.png", a.Position())
    }
}

func main() {
    InitCanvas(screen)
    MainLoop(tick)
}

